src/data/player.py

```diff
@@ -643,6 +528,318 @@ def set_board(self, board_id: int, board_number: int, color: BoardColor):
+                continue
+            if (
+                p.federation == Federation(self.tournament.event.federation)
+                or p.federation == 'NON'  # 1.4.2a
+            ):
+                continue
+            missed_rounds = 0
+            for r, pairing in p.pairings_by_round.items():
+                if pairing.unplayed and pairing.result not in [
+                    Result.PAIRING_ALLOCATED_BYE,
+                    Result.REST_GAME,
+                ]:
+                    missed_rounds += 1
+            if missed_rounds > 1:
+                continue
+
+            eligible_players.append(p)
+
+        # Per-round counts among eligible & present -----------------
+        worst_players: float = float('inf')
+        worst_federations: float = float('inf')
+        worst_titled: float = float('inf')
+
+        for rnd in range(1, self.tournament.rounds + 1):
+            present: list[Player] = []
+            for p in eligible_players:
+                pairing: Pairing | None = p.pairings_by_round.get(rnd)
+                if pairing and (
+                    pairing.played
+                    or pairing.result
+                    in [
+                        Result.PAIRING_ALLOCATED_BYE,
+                        Result.REST_GAME,
+                    ]
+                ):
+                    present.append(p)
+
+            n_players = len(present)
+            n_titled = sum(1 for p in present if p.title != PlayerTitle.NONE)
+
+            # 1.4.2a
+            present_not_fid = [p for p in present if p.federation != 'FID']
+            n_feds = len({p.federation for p in present_not_fid})
+
+            # Track worst (minimum) across rounds
+            worst_players = min(worst_players, n_players)
+            worst_federations = min(worst_federations, n_feds)
+            worst_titled = min(worst_titled, n_titled)
+
+        # Handle case of zero rounds gracefully
+        if worst_players is float('inf'):
+            worst_players = 0
+            worst_federations = 0
+            worst_titled = 0
+
+        msg = _(
+            '<b>1.4.3d</b> Swiss System tournaments in which participants include in every round at least 20 FIDE rated players, not from the host federation, from at least 3 different federations, at least 10 of whom hold GM, IM, WGM or WIM titles.'
+        )
+        for tn, res in results.items():
+            res.all_federations_count = int(worst_federations)
+            res.not_enough_all_federations = (
+                msg if res.all_federations_count < 3 else None
+            )
+
+            res.eligible_players_title_count = int(worst_titled)
+            res.not_enough_all_title_holders = (
+                msg if res.eligible_players_title_count < 10 else None
+            )
+
+            res.eligible_players_count = int(worst_players)
+            res.not_enough_foreign_players = (
+                msg if res.eligible_players_count < 20 else None
+            )
```

Comment:
Hard to know where to put the comment about 1.4.3.d
I think all code related to that should ideally be in a tournament `cached_property`